### ISSUE TITLE
Latest 5 alpha testing feedback

### ISSUE THREAD

**Issue by mfendeksilverstripe** [tractorcow-farm/silverstripe-fluent#621]
# Feedback for 5.0.0-alpha7

I'm providing some feedback on the latest alpha version.

### Locales

* NZ global default
* US - fallback to NZ
* AU - fallback to NZ
* JP - no fallback

### Pages

Test page localised in NZ and AU.

### Terminology

**Unlocalise** - should be used when we archive a locale and there is at least one other locale left
**Archive** - should be used to indicate that the last locale is being archived

It's really important to make clear to content author the distinction between these two actions as **Archive** is potentially much more destructive than the **Unlocalise**.

#### Localise widget UI

![Screen Shot 2020-06-05 at 11 54 14 AM](https://user-images.githubusercontent.com/26395487/83821818-811bd400-a723-11ea-9dd1-fb7ba7e844c3.png)

`Clear from all except current locale` should be changed to `Unlocalise page in all other locales`.
`Copy current locale to other locales` should be changed to `Localise page in all other locales using the content of this page`.

### More actions widget

This widget doesn't work well with Fluent as it's reading data from the base record instead of the localised record.

![Screen Shot 2020-06-05 at 12 31 03 PM](https://user-images.githubusercontent.com/26395487/83823613-80d20780-a728-11ea-8886-ec48ac24c4bd.png)

Base record always contains data from the most recently updated locale. I understand the technical reason for it but from the point of content author this seems really random and misleading.

For example in my test setup I have a page which is unpublished in NZ and later published in AU. If you go to the JP which does't have localised content nor it inherits content it will still show Last edited and publish dates from the AU locale as the that data is on the base record. You get the same result on the US locale.

Additionally, some of the actions should not be even displayed for this page such as `Cancel draft changes` and `Unpublish`

There are several option on how to tackle this. For example:

* apply extension after the vanilla fields are created and update them as needed
* override `isPublished` and other methods (this would require some refactoring of the `Versioned` module)

### How to localise a page

Currently, there are 3 ways how to create / publish a page localisation. This is a bit confusing as these options overlap. A simpler more consistent UI is preferred. First, let's have a look at current state.

#### Option 1 - Localise by action buttons

![Screen Shot 2020-06-05 at 12 09 04 PM](https://user-images.githubusercontent.com/26395487/83822517-5df22400-a725-11ea-9887-1d5c261864be.png)

#### Option 2 - Batch actions

![Screen Shot 2020-06-05 at 12 10 21 PM](https://user-images.githubusercontent.com/26395487/83822583-8bd76880-a725-11ea-9297-67b01bfa5605.png)

#### Option 3 - Localisation manager

![Screen Shot 2020-06-05 at 12 11 10 PM](https://user-images.githubusercontent.com/26395487/83822624-a90c3700-a725-11ea-96ed-be3d78e0d565.png)

There are some issues with this approach:

* content authors prefer to have one way how to do one specific thing, they prefer simplicity
* Option 1 produces random (from content author point of view) results when used in a locale which inherits directly from base record
* Option 2 is potentially really destructive for example clear all other locales will unlocalise potentially a lot of content and the publish all potentially adds all locales to the filtered locales list

#### Recommendation

* remove Option 1 buttons completely as these actions can be achieved via Option 3 and that way it's much more clear what is going on
* in case page doesn't have it's own content (it either inherits from another locale or the base record) switch the page edit form to read only but keep the Localisation manager available
* keep the Option 2 but move it out of the main actions area
* make Options 3 the main / recommended way how to manage localisations for a page

Overall, I'm really happy with the features of this module, I especially like the Localisation manager tab.

Please let me know what you think.

**Comment by tractorcow** [comment]
Hi @mfendeksilverstripe I first want to say thanks for the great feedback. It's given in great detail and very precise; It's going to be helpful for the final release.

I think changing "Clear from" to "Unlocalise" is a great idea. It provides a good inverse to "localise". :)

I do think we need to be careful about too long sentences; Here is a suggested change.

`Localise page in all other locales using the content of this page.` could be 'Localise current content to all locales`? (note: Not all records are pages, so I'm trying to be agnostic between the object types).

> This widget doesn't work well with Fluent as it's reading data from the base record instead of the localised record
... (snip)
>  it will still show Last edited and publish dates from the AU locale as the that data is on the base record. 

Yeah, I am stuck on overriding the core version functions... as you've noted, a lot of these methods are not overridable. I half considered overriding versioned entirely, and injecting a replacement, but that's a big and risky change to make.

A bigger issue is the versioned admin extension; the whole list is incorrect based on which locale you look at. I've been trying to figure out how to tackle these issues at once, so I'm trying to be careful not to monkey patch small symptoms too much without fixing the core issues. (i.e. those versioned methods).

> Option 1 produces random (from content author point of view) results when used in a locale which inherits directly from base record

I agree. I would suggest removing this option entirely, and forcing the user to select a version to either:
 - modify the form (data source is entered text)
 - explicitly select a source record (i.e. an already localised source to use as the base).

I agree that the "base" record is not a good source, so I'd like to de-prioritise it as a source, even if that means disabling the "save" button.

An issue with option 2 is that "save and publish (all locales)" probably should not be enabled if you are viewing the record in an un-localised locale. It has the same issue as "copy to draft"; Incorrectly using the ambiguous base record.



So, yes you have correctly identified a lot of the UX issues with the module, although fixing them is a challenge. :)

**Comment by tractorcow** [comment]
Immediate actions to take:
 - Provide protection for users saving FROM an unlocalised locale. This could either be a hard-protection (disable some actions), or a soft protection (display conditional warnings).
 - Do some text / action renaming to tighten up terminology.
- Categorise versioned API surface area that is reporting incorrectly (e.g. isPublished(), published date, etc). Ideally raise this as a versioned module issue and seek feedback from the community.

**Comment by tractorcow** [comment]
and by immediate I mean "whenever I feel like it, maybe tomorrow or months from now"... haha... I will make an effort though. :)

**Comment by tractorcow** [comment]
Here's an idea to minimise the impact of the first issue:

If a record isn't localised, we could still use the base record to display some fields, but we should probably whitelist these, instead of copying the whole record.

The main reason for requiring a base record is that there were cases the CMS was unusable after changing locales. (e.g. the entire pages tree would be a list of empty nodes).

In order to minimise garbage data being accidentally promoted to new localisations, we could limit say, Title, URLSegment, MenuTitle only... (just enough to create a findable but empty page) and require the user to manually enter these in.

This would only occur if the fallback did not have records of course. :) You'd get the fallback content in that case.

Example use case;

```
class FluentSiteTree {
    /**
     * If editing an unlocalised record, only these can be inherited
     * from the "base" record saved in the last locales
     */
    private static $whitelist_base_fields = [
        'Title',
        'URLSegment',
    ];
}
```

**Comment by t3hn0** [comment]
Great post! I think it's time to share our approach with base record problem too.

We thought that best solution in our case would be to allow population of certain base fields only with base locale values and skip others.
I achieved that by modifying function localiseManipulationTable in FluentExtension.php. I added this before the end of function:

`
        if( $this->owner instanceof SiteTree && $this->owner->exists() && !$locale->IsGlobalDefault ) {
            $skip_fields = [
                'Title',
                'MenuTitle',
                'URLSegment',
                'Content',
                'MetaDescription',
                'ExtraMeta',
            ];
            
            if( isset($manipulation['SiteTree']) ) {
                foreach( $skip_fields as $field ) {
                    if( isset($manipulation['SiteTree']['fields'][$field]) ) {
                        unset($manipulation['SiteTree']['fields'][$field]);
                    }
                }
            }
            if( isset($manipulation['SiteTree_Live']) ) {
                foreach( $skip_fields as $field ) {
                    if( isset($manipulation['SiteTree']['fields'][$field]) ) {
                        unset($manipulation['SiteTree']['fields'][$field]);
                    }
                }
            }
        }`

It did the trick. Versions table didn't matter at the moment.

**Comment by tractorcow** [comment]
That's also an interesting approach, but won't that just mean that the base record gets the value that was first written, instead of the last? (similar, but different issue).

**Comment by t3hn0** [comment]
It will allow to write data to base table if record is beeing created in "foreign" locale (to avoid empty rows in base table) but it will allow to overwrite those rows only in default locale. In our case pages are beeing created in default locale by default, so it suits our need.

And another thing I forgot to mention earlier about benefit of leaving default locale translations in base table is for example if someone would remove Fluent extension later on and would be left with page in default locale.

**Comment by tractorcow** [comment]
I see, so you would have (for example) "New Blog Page" etc essentially permanently in the base table, but no other fields?

**Comment by tractorcow** [comment]
> And another thing I forgot to mention earlier about benefit of leaving default locale translations in base table is for example if someone would remove Fluent extension later on and would be left with page in default locale.

The problem with this approach that I've found is that sometimes records are created, but never saved in the default locale, so that isn't always possible. In fluent 4 and above I had to de-prioritise the concept of "default" locale, and instead look at fallback chains on query.

I do quite like the idea of limiting base record to the initial write, though (maybe) so I might explore your solution a bit more and see how that can work in the core module. :) Maybe as an opt in feature?
